src/chart/sparkline.js

```diff
@@ -0,0 +1,117 @@
+(function(d3, fc) {
+    'use strict';
+
+    fc.charts.sparkline = function() {
+
+        // creates an array with four elements, representing the high, low, open and close
+        // values of the given array
+        function highLowOpenClose(data) {
+            var xValueAccessor = sparkline.xValue(),
+                yValueAccessor = sparkline.yValue();
+
+            var high = d3.max(data, yValueAccessor);
+            var low = d3.min(data, yValueAccessor);
+
+            function elementWithYValue(value) {
+                return data.filter(function(d) {
+                    return yValueAccessor(d) === value;
+                })[0];
+            }
+
+            return [{
+                    x: xValueAccessor(data[0]),
+                    y: yValueAccessor(data[0])
+                }, {
+                    x: xValueAccessor(elementWithYValue(high)),
+                    y: high
+                }, {
+                    x: xValueAccessor(elementWithYValue(low)),
+                    y: low
+                }, {
+                    x: xValueAccessor(data[data.length - 1]),
+                    y: yValueAccessor(data[data.length - 1])
+                }];
+        }
+
+        var xScale = fc.scale.dateTime();
+        var yScale = d3.scale.linear();
+        var line = fc.series.line();
+
+        // configure the point series to render the data from the
+        // highLowOpenClose function
+        var point = fc.series.point()
+            .xValue(function(d) { return d.x; })
+            .yValue(function(d) { return d.y; })
+            .decorate(function(sel) {
+                sel.attr('class', function(d, i) {
+                    switch (i) {
+                        case 0: return 'open';
+                        case 1: return 'high';
+                        case 2: return 'low';
+                        case 3: return 'close';
+                    }
+                });
+            });
+
+        var multi = fc.series.multi()
+            .series([line, point])
+            .mapping(function(data, series) {
+                switch (series) {
+                    case point:
+                        return highLowOpenClose(data);
+                    default:
+                        return data;
+                }
+            });
+
+        var sparkline = function(selection) {
+
+            point.radius(sparkline.radius.value);
+
+            selection.each(function(data) {
+
+                var container = d3.select(this);
+
+                var mainContainer = container.selectAll('g')
+                    .data([data]);
+                mainContainer.enter()
+                    .append('g')
+                    .classed('sparkline', true)
+                    .layout({
```

Comment:
Could we not just pad the range (pixel) values by the radius value?

On Thu, 7 May 2015 12:33 Colin Eberhardt <notifications@github.com> wrote:

> In src/chart/sparkline.js
> <https://github.com/ScottLogic/d3-financial-components/pull/338#discussion_r29843724>
> :
>
> > +            });
> > +
> > +        var sparkline = function(selection) {
> > +
> > +            point.radius(sparkline.radius.value);
> > +
> > +            selection.each(function(data) {
> > +
> > +                var container = d3.select(this);
> > +
> > +                var mainContainer = container.selectAll('g')
> > +                    .data([data]);
> > +                mainContainer.enter()
> > +                    .append('g')
> > +                    .classed('sparkline', true)
> > +                    .layout({
>
>  As in could we just pad the ranges?
>
> Not easily - if I understand you correctly!
>
> The margin around the container is in the screen coordinate system, it is
> the margin required to ensure that the circles do not fall outside of the
> bounds on the container element.
>
> To achieve this via the scale ranges would require converting this into
> the domain coordinate system.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3-financial-components/pull/338/files#r29843724>
> .
>